### PR TITLE
ENH: Added proj_context_clone

### LINKED ISSUE #2382: PROPOSAL: Add proj_context_clone
Reasoning:
1, To buffer a library from the potential setting changes made by other libraries on the global context (even though not recommended to change global context settings, they can still do this).
2. To set default context settings on an application global context that will be applied to any new context.

With the `proj_context_clone` method, a library can create their application level global context with `proj_context_create` and then apply any settings they would like. Then, when creating new contexts, they can use `proj_context_clone` to create new contexts.

### PR DESCRIPTION
**Pull request by snowman2**
- [x] Closes #2382
 - [x] Tests added
 - [x] Added clear title that can be used to generate release notes
 - [x] Fully documented, including updating `docs/source/*.rst` for new API

### CODE REVIEW

File: src/4D_api.cpp [hunk 1]
```diff
@@ -1448,11 +1448,18 @@ int proj_errno_reset (const PJ *P) {
...
 PJ_CONTEXT *proj_context_create (void) {
     return pj_ctx_alloc ();
 }
 
+/* Create a new context based on a custom context */
+PJ_CONTEXT *proj_context_clone (PJ_CONTEXT *ctx) {
+    if (nullptr==ctx)
+        return pj_ctx_alloc ();
+
+    return pj_ctx_clone (ctx);
```
Comment (rouault) [inline src/4D_api.cpp]: this function should be moved to ctx.cpp() and merged with it

File: src/ctx.cpp [hunk 2]
```diff
@@ -225,6 +225,16 @@ projCtx pj_ctx_alloc()
     return new (std::nothrow) projCtx_t(*pj_get_default_ctx());
 }
 
+/************************************************************************/
+/*                            pj_ctx_clone()                            */
+/************************************************************************/
+
+projCtx pj_ctx_clone( projCtx ctx )
```
Comment (rouault) [inline src/ctx.cpp]: we don't need that function

File: src/proj_api.h [hunk 3]
```diff
@@ -192,6 +192,7 @@ void PROJ_DLL pj_cleanup_lock(void);
 
 void PROJ_DLL pj_set_ctx( projPJ, projCtx );
 projCtx PROJ_DLL pj_ctx_alloc(void);
+projCtx PROJ_DLL pj_ctx_clone( projCtx );
```
Comment (rouault) [inline src/proj_api.h]: to remove. proj_api.h is legacy and is going to be removed soon

File: src/proj_symbol_rename.h [hunk 4]
```diff
@@ -29,6 +29,7 @@
 #define pj_clear_initcache internal_pj_clear_initcache
 #define pj_compare_datums internal_pj_compare_datums
 #define pj_ctx_alloc internal_pj_ctx_alloc
+#define pj_ctx_clone internal_pj_ctx_clone
```
Comment (rouault) [inline src/proj_symbol_rename.h]: to remove

File: scripts/reference_exported_symbols.txt [hunk 5]
```diff
@@ -752,6 +752,7 @@ pj_compare_datums
 pj_context_get_grid_cache_filename(projCtx_t*)
 pj_context_set_user_writable_directory(projCtx_t*, std::string const&)
 pj_ctx_alloc
+pj_ctx_clone
```
Comment (rouault) [inline scripts/reference_exported_symbols.txt]: to remove

File: src/4D_api.cpp (hunk 1 above)
Comment (snowman2) [reply src/4D_api.cpp]: :+1:

File: src/ctx.cpp (hunk 2 above)
Comment (snowman2) [reply src/ctx.cpp]: :+1:

File: src/proj_api.h (hunk 3 above)
Comment (snowman2) [reply src/proj_api.h]: :+1:

**Comment by snowman2** [comment]
Are there any remaining changes needed to get this into 7.2? The release is coming soon and it would be nice to have it added :crossed_fingers:

**Comment by snowman2** [comment]
Thanks @rouault :+1:

**Pull request merged**